### ISSUE TITLE
Attribute macro can replace item with expression.

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#144716]
I tried this code:

bar/src/lib.rs:
```rust
use proc_macro::TokenStream;

#[proc_macro_attribute]
pub fn my_macro(_: TokenStream, _: TokenStream) -> TokenStream {
    "\"weird\"".parse().unwrap()
}
```

src/main.rs:
```rust
fn main() {
    // prints "weird"
    println!("{}", {
        #[bar::my_macro]
        struct Thing;
    });
}
```

I expected the code to give a compile error. Instead, it compiles and prints `weird`.

The proc macro is somehow replacing an item with an expression. The [reference](https://doc.rust-lang.org/reference/procedural-macros.html#r-macro.proc.attribute.def) says this isn't allowed:
> The returned `TokenStream` replaces the `item` with an arbitrary number of `items`.

Initially discovered in  #144579.

@rustbot labels +A-macros +A-proc-macros +A-attributes

### Meta

`rustc --version --verbose`:
```
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: aarch64-apple-darwin
release: 1.88.0
LLVM version: 20.1.5
```

**Comment by danielhenrymantilla** [comment]
This might be an issue with the reference rather than a bug in implementation?

  - One example where this came in handy: https://docs.rs/cfg_eval/0.1.2/cfg_eval/index.html#example

**Comment by Noratrieb** [comment]
@petrochenkov about your opinion

**Comment by petrochenkov** [comment]
The macro invocation is in a statement position, so it can produce multiple statements, including expression statements.
So it's not necessarily an item.

**Comment by theemathas** [comment]
Expression statements do not produce values though, right? If the macro expanded to an expression statement (as opposed to a non-statement expression), I would not have been able to print the value `"weird"`.

**Comment by danielhenrymantilla** [comment]
They can when within braces:
```rs
{
    foo! {} // say this emits `42`
    bar! {} // and this, ``
} // <- evaluates to `42`
```

  - [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=3779c78d5922f4ef839a28e5f0c61be1)

**Comment by theemathas** [comment]
And..... the reference says that the result of an expression statement [is always ignored](https://doc.rust-lang.org/stable/reference/statements.html#r-statement.expr.intro).

This function compiles, and returns 42:
```rust
macro_rules! nothing {
    () => {}
}

fn demo() -> i32 {
    { 42 }
    nothing! {}
}
```

Based on the grammar in the reference, I believe that the `{ 42 }` must be an expression statement. However, the result of it is clearly not ignored.

Either the implementation or the reference is also wrong here.

**Comment by danielhenrymantilla** [comment]
Yes, I believe the reference ought to mention the special case of the without-semicolon "block statement", whose value is not necessarily always discarded. Do you want to open an issue about it?

**Comment by kpreid** [comment]
> This function compiles, and returns 42:

IMO, this _shouldn't_ work. It appears to violate the tree structure of the language in a way that macros usually can't (e.g. `foo(a, b!(), c)` always has 3 arguments), and thus surprisingly reduces the ability to read Rust without knowing everything about every involved macro expansion.

**Comment by petrochenkov** [comment]
> (e.g. `foo(a, b!(), c)` always has 3 arguments

What about `foo(a, #[my_cfg(false)] b, c)`, how many arguments does it have?
